pairing/plonk/src/prover.rs

```diff
@@ -0,0 +1,293 @@
+            map_or(Err(ProofError::PolyDivision), |s| Ok(s))?;
+        if r.is_zero() == false {return Err(ProofError::PolyDivision)}
+
+        // split t to fit to the commitment
+        let tlow: DensePolynomial<E::Fr>;
+        let mut tmid = DensePolynomial::from_coefficients_slice(&[E::Fr::zero()]);
+        let mut thgh = DensePolynomial::from_coefficients_slice(&[E::Fr::zero()]);
+        if t.coeffs.len() <= n {tlow = t}
+        else if t.coeffs.len() <= 2*n
+        {
+            tlow = DensePolynomial::from_coefficients_slice(&t.coeffs[0..n]);
+            tmid = DensePolynomial::from_coefficients_slice(&t.coeffs[n..t.coeffs.len()]);
+        }
+        else
+        {
+            tlow = DensePolynomial::from_coefficients_slice(&t.coeffs[0..n]);
+            tmid = DensePolynomial::from_coefficients_slice(&t.coeffs[n..2*n]);
+            thgh = DensePolynomial::from_coefficients_slice(&t.coeffs[2*n..]);
+        }
+
+        // commit to tlow, tmid, thgh
+        let tlow_comm = index.urs.get_ref().commit(&tlow)?;
+        let tmid_comm = index.urs.get_ref().commit(&tmid)?;
+        let thgh_comm = index.urs.get_ref().commit(&thgh)?;
+
+        // absorb the polycommitments into the argument and sample zeta
+        fq_sponge.absorb_g(&[tlow_comm, tmid_comm, thgh_comm]);
+        oracles.zeta = fq_sponge.challenge();
+        let zeta2 = oracles.zeta.pow(&[n as u64]);
+        let zeta3 = zeta2.square();
+
+        // compute linearisation polynomial
+
+        evals.a = a.evaluate(oracles.zeta);
+        evals.b = b.evaluate(oracles.zeta);
+        evals.c = c.evaluate(oracles.zeta);
+        evals.sigma1 = index.cs.sigmam[0].evaluate(oracles.zeta);
+        evals.sigma2 = index.cs.sigmam[1].evaluate(oracles.zeta);
+        evals.z = z.evaluate(oracles.zeta * &index.cs.domain.group_gen);
+
+        let bz = oracles.beta * &oracles.zeta;
+        let r1 =
+            &(&(&(&index.cs.qm.scale(evals.a*&evals.b) +
+            &index.cs.ql.scale(evals.a)) +
+            &index.cs.qr.scale(evals.b)) +
+            &index.cs.qo.scale(evals.c)) +
+            &index.cs.qc;
+        let r2 =
+            z.scale
+            (
+                (evals.a + &bz + &oracles.gamma) *
+                &(evals.b + &(bz * &index.cs.r) + &oracles.gamma) *
+                &(evals.c + &(bz * &index.cs.o) + &oracles.gamma) *
+                &oracles.alpha
+            );
+        let r3 =
+            index.cs.sigmam[2].scale
+            (
+                (evals.a + &(oracles.beta * &evals.sigma1) + &oracles.gamma) *
+                &(evals.b + &(oracles.beta * &evals.sigma2) + &oracles.gamma) *
+                &(oracles.beta * &evals.z * &oracles.alpha)
+            );
+        let r4 = z.scale(alpsq * &index.cs.l0.evaluate(oracles.zeta));
+        let r = &(&(&r1 + &r2) - &r3) + &r4;
+        evals.r = r.evaluate(oracles.zeta);
+
+        // query opening scaler challenge
```

Comment:
Should be `scalar`

Reply:
Scaler deriving from "to scale" with DensePolynomial::scale function.